### PR TITLE
Add test for `deferPhase` in blueprint actions

### PR DESCRIPTION
**Pull request by viveksinghggits**
## Change Overview

As part of [this PR](https://github.com/kanisterio/kanister/pull/1297) we tried to add an eventual phase in the kanister blueprint action.
This PR adds test for that change.

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [x] :robot: Test

## Issues

- #XXX

## Test Plan

- [ ] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

```
» go test -check.f "ControllerSuite"
```
The logs can be found [here](https://gist.github.com/d66ccb95105eccb47f0107a18d19ae6b).

### CODE REVIEW

**Comment by viveksinghggits** [comment]
@ihcsim this can also be reviewed.

File: pkg/controller/controller_test.go [hunk 1]
```diff
@@ -189,6 +224,158 @@ func newBPWithOutputArtifact() *crv1alpha1.Blueprint {
...
+	return &crv1alpha1.BlueprintPhase{
+		Name: "deferPhase",
+		Func: function.KubeTaskFuncName,
+		Args: map[string]interface{}{
+			"image":     mysqlSidecarImage,
+			"namespace": "default",
+			"command":   []string{"kando", "output", "value", "deferValue"},
+		},
+	}
+}
```
Comment (ihcsim) [inline pkg/controller/controller_test.go:323]: Feels like this is almost the same as `backupPhase()`. Can we just change the `backupPhase()` signature to something like:

```go
func backupPhase(name string, command []string) crv1alpha1.BlueprintPhase
```

File: pkg/controller/controller_test.go [hunk 2]
```diff
@@ -498,6 +685,123 @@ func (s *ControllerSuite) TestRuntimeObjEventLogs(c *C) {
 	c.Assert(len(events.Items), Equals, 0)
 }
 
+func (s *ControllerSuite) TestDeferPhase(c *C) {
+	os.Setenv(kube.PodNSEnvVar, "test")
```
Comment (ihcsim) [inline pkg/controller/controller_test.go:689]: For good test hygiene, may want to consider unsetting this env var using Go `defer` after the test is done.

File: pkg/controller/controller_test.go [hunk 3]
```diff
@@ -189,6 +224,158 @@ func newBPWithOutputArtifact() *crv1alpha1.Blueprint {
...
+		Func: function.KubeTaskFuncName,
+		Args: map[string]interface{}{
+			"image":     mysqlSidecarImage,
+			"namespace": "default",
+			"command":   []string{"kando", "output", "value", "deferValue"},
+		},
+	}
+}
+
+func newBPWithDeferPhase() *crv1alpha1.Blueprint {
```
Comment (ihcsim) [inline pkg/controller/controller_test.go:325]: Ideally, we should re-use the `NewTestBlueprint()` constructor in the `testutil` package. But the current implementation in `testutil` looks a bit rigid. Maybe in the future we can add better builder interfaces to them.

E.g.,
```go
package testutil

 // without any embedded default action
func NewTestBlueprint() *crv1alpha1.Blueprint

func BlueprintWithAction(blueprint *crv1alpha1.Blueprint, action *crv1alpha1.BlueprintAction) *crv1alpha1.Blueprint

func ActionWithArtifact(action *crv1alpha1.BlueprintAction, artifact *crv1alpha1.Artifact) *crv1alpha1.BlueprintAction

func ActionWithPhase(blueprint *crv1alpha1.Blueprint , phase *crv1alpha1.BlueprintPhase) *crv1alpha1.BlueprintAction 
```

**Review by ihcsim — APPROVED** [review]
LGTM - thanks for adding tests!

File: pkg/controller/controller_test.go (hunk 3 above)
Comment (viveksinghggits) [reply pkg/controller/controller_test.go:300]: yeah, I agree we should track and work on this.

**Review by ihcsim — APPROVED** [review]

**Pull request merged**